src/compiler/postprocess.rs

```diff
@@ -0,0 +1,98 @@
+use std::io::{BufWriter, Write};
+use std::path::PathBuf;
+use std::process::{Command, Stdio};
+use std::sync::{Arc, Mutex, MutexGuard};
+
+use itertools::Itertools;
+use wildmatch::WildMatchPattern;
+
+#[derive(Debug, Clone)]
+pub struct PostProcessor {
+    pub name: String,
+    pub bin: PathBuf,
+
+    command: Arc<Mutex<Command>>
+}
+
+impl PostProcessor {
+    pub fn new<N: Into<String>, B: Into<PathBuf>>(name: N, bin: B) -> Self {
+        let name: String = name.into();
+        let bin: PathBuf = bin.into();
+        let mut command = Command::new(bin.clone());
+        command.stdin(Stdio::piped());
+        command.stdout(Stdio::piped());
+        command.stderr(Stdio::piped());
+        let command = Arc::new(Mutex::new(command));
+        Self {
+            name,
+            bin,
+            command,
+        }
+    }
+
+    pub fn cmd(&self) -> MutexGuard<Command> {
+        self.command.lock().expect("Couldn't lock on command (arc)")
+    }
+
+    pub fn is_available(&self) -> bool {
+        match Command::new(self.bin.clone()).spawn() {
+            Ok(mut v) => {
+                let _ = v.kill();
+                true
+            },
+            Err(_) => false
+        }
+    }
+
+    pub fn execute(&self, code: String) -> Result<String, Box<dyn std::error::Error>> {
+
+        if !self.is_available() { return Ok(code) }
+
+        let mut spawned = self.cmd().spawn()?;
+        
+        // send to stdin
+        {
+            if let Some(stdin) = spawned.stdin.as_mut() {
+                let mut writer = BufWriter::new(stdin);
+                writer.write_all(code.as_bytes())?;
+                writer.flush()?;
+            } else {
+                return Err(String::new().into())
+            }
+        }
+
+        // read from stdout
+        {
+            let res = spawned.wait_with_output()?;
+            Ok(String::from_utf8(res.stdout)?)
+        }
+    }
+
+    pub fn get_default() -> Vec<Self> {
+        let mut postprocessors = Vec::new();
+        
+        let shfmt = PostProcessor::new("shfmt", "/usr/bin/shfmt");
```

Comment:
Here you're creating both processors, and then immediately removing them according to the CLI options; this is inelegant.  Surely it would be better to combine `get_default()` and `filter_default()`, and only create the processors you're actually going to need:
```
pub fn get_processors(filters: Vec<WildMatchPattern<'*', '?'>>) -> Vec<Self> {
    let mut processors = Vec::new();
    if Self::filter_processor(&filters, "shfmt") {
        // Add "shfmt"
    }
    if Self::filter_processor(&filters, "bshchk") {
        // Add "bshchk"
    }
    processors
}
```
I suspect you did it this way so you could call `get_default()` in the integration tests, but you could always call this new function with an empty vector:
```
let default = PostProcessor::get_processors(vec![]);
```

Reply:
> Surely it would be better to combine `get_default()` and `filter_default()`, and only create the processors you're actually going to need:

why? its still two functions but now it breaks DRY and a little bit more dirty. there are no issues with the current setup